demos/multi_material/compositional_buoyancy/van_keken_1997.py

```diff
@@ -0,0 +1,276 @@
+# Group parameters defining the cosine profile
+isd_params = (interface_deflection, interface_wavelength, material_interface_y)
+
+# Shapely LineString representation of the material interface
+interface_x = np.linspace(0, lx, 1000)
+interface_y = cosine_curve(interface_x, *isd_params)
+line_string = sl.LineString([*np.column_stack((interface_x, interface_y))])
+sl.prepare(line_string)
+
+# Extract node coordinates
+node_coords_x, node_coords_y = node_coordinates(psi)
+# Determine to which material nodes belong and calculate distance to interface
+node_relation_to_curve = [
+    (
+        node_coord_y > cosine_curve(node_coord_x, *isd_params),
+        line_string.distance(sl.Point(node_coord_x, node_coord_y)),
+    )
+    for node_coord_x, node_coord_y in zip(node_coords_x, node_coords_y)
+]
+
+# Define the signed-distance function and overwrite its value array
+signed_dist_to_interface = Function(K)
+signed_dist_to_interface.dat.data[:] = [
+    dist if is_above else -dist for is_above, dist in node_relation_to_curve
+]
+
+# Define thickness of the hyperbolic tangent profile
+min_mesh_edge_length = min(lx / nx, ly / ny)
+epsilon = Constant(min_mesh_edge_length / 4)
+
+# Initialise level set as a smooth step function
+psi.interpolate((1 + tanh(signed_dist_to_interface / 2 / epsilon)) / 2)
+# -
+
+# We next define materials present in the simulation. Here, the problem is
+# non-dimensionalised and can be described by the product of the expressions for the
+# Rayleigh and buoyancy numbers, RaB, which is also referred to as compositional
+# Rayleigh number. Therefore, we provide a value for thermal and compositional Rayleigh
+# numbers to define our approximation.
+
+# +
+buoyant_material = Material(RaB=-1)  # Vertical direction is flipped in the benchmark
+dense_material = Material(RaB=0)
+materials = [buoyant_material, dense_material]
+
+Ra = 0  # Thermal Rayleigh number
+
+RaB = field_interface(
+    [psi], [material.RaB for material in materials], method="arithmetic"
+)  # Compositional Rayleigh number, defined based on each material value and location
+
+approximation = BoussinesqApproximation(Ra, RaB=RaB)
+# -
+
+# As with the previous examples, we set up a *Timestep Adaptor* for controlling the
+# time-step length (via a CFL criterion) as the simulation advances in time. We specify
+# the initial time, initial time step $\Delta t$, and output frequency (in time units).
+
+time_now = 0  # Initial time
+delta_t = Function(R).assign(1)  # Initial time step
+output_frequency = 10  # Frequency (based on simulation time) at which to output
+t_adapt = TimestepAdaptor(
+    delta_t, u, V, target_cfl=0.6, maximum_timestep=output_frequency
+)
+
+# This problem has a constant pressure nullspace, handled identically to our previous
+# tutorials.
+
+Z_nullspace = create_stokes_nullspace(Z)
+
+# Boundary conditions are specified next: no slip at the top and bottom and free slip
```

Comment:
Is it worth adding a comment here that since all domain boundaries are closed, there is no boundary condition for the level set required?